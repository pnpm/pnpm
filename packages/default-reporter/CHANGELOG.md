# @pnpm/default-reporter

## 7.9.1

### Patch Changes

- 3b8e3b6b1: Always print the final progress stats.
- Updated dependencies [767212f4e]
- Updated dependencies [092f8dd83]
  - @pnpm/config@11.3.0

## 7.9.0

### Minor Changes

- 663afd68e: Scope is not reported when the scope is only one project.
- 86cd72de3: Show the progress of adding packages to the virtual store.

### Patch Changes

- Updated dependencies [86cd72de3]
  - @pnpm/core-loggers@5.0.0

## 7.8.0

### Minor Changes

- 09b42d3ab: Use RxJS instead of "most".

## 7.7.0

### Minor Changes

- af8361946: Sometimes, when installing new dependencies that rely on many peer dependencies, or when running installation on a huge monorepo, there will be hundreds or thousands of warnings. Printing many messages to the terminal is expensive and reduces speed, so pnpm will only print a few warnings and report the total number of the unprinted warnings.

## 7.6.4

### Patch Changes

- Updated dependencies [75a36deba]
- Updated dependencies [9f1a29ff9]
  - @pnpm/error@1.3.1
  - @pnpm/config@11.2.7

## 7.6.3

### Patch Changes

- 13c332e69: Fixes a regression published in pnpm v5.5.3 as a result of nullish coalescing refactoring.

## 7.6.2

### Patch Changes

- Updated dependencies [ac0d3e122]
  - @pnpm/config@11.2.6

## 7.6.1

### Patch Changes

- Updated dependencies [972864e0d]
  - @pnpm/config@11.2.5

## 7.6.0

### Minor Changes

- 6d480dd7a: Print the authorization settings (with hidden private info), when an authorization error happens during fetch.

### Patch Changes

- Updated dependencies [6d480dd7a]
  - @pnpm/error@1.3.0
  - @pnpm/config@11.2.4

## 7.5.4

### Patch Changes

- Updated dependencies [13c18e397]
  - @pnpm/config@11.2.3

## 7.5.3

### Patch Changes

- Updated dependencies [3f6d35997]
  - @pnpm/config@11.2.2

## 7.5.2

### Patch Changes

- Updated dependencies [a2ef8084f]
  - @pnpm/config@11.2.1

## 7.5.1

### Patch Changes

- Updated dependencies [ad69677a7]
  - @pnpm/config@11.2.0

## 7.5.0

### Minor Changes

- 9a908bc07: Print info after install about hardlinked/copied packages in `node_modules/.pnpm`

### Patch Changes

- Updated dependencies [9a908bc07]
- Updated dependencies [9a908bc07]
  - @pnpm/core-loggers@4.2.0

## 7.4.7

### Patch Changes

- Updated dependencies [65b4d07ca]
- Updated dependencies [ab3b8f51d]
  - @pnpm/config@11.1.0

## 7.4.6

### Patch Changes

- @pnpm/config@11.0.1

## 7.4.5

### Patch Changes

- Updated dependencies [71aeb9a38]
- Updated dependencies [915828b46]
  - @pnpm/config@11.0.0

## 7.4.4

### Patch Changes

- @pnpm/config@10.0.1

## 7.4.3

### Patch Changes

- 220896511: Remove common-tags from dependencies.
- Updated dependencies [db17f6f7b]
- Updated dependencies [1146b76d2]
- Updated dependencies [db17f6f7b]
  - @pnpm/config@10.0.0
  - @pnpm/types@6.2.0
  - @pnpm/core-loggers@4.1.2

## 7.4.2

### Patch Changes

- Updated dependencies [71a8c8ce3]
- Updated dependencies [71a8c8ce3]
  - @pnpm/types@6.1.0
  - @pnpm/config@9.2.0
  - @pnpm/core-loggers@4.1.1

## 7.4.1

### Patch Changes

- e934b1a48: Update chalk to v4.1.0.

## 7.4.0

### Minor Changes

- 2ebb7af33: New reporter added for request retries.

### Patch Changes

- Updated dependencies [2ebb7af33]
  - @pnpm/core-loggers@4.1.0

## 7.3.0

### Minor Changes

- eb82084e1: Color the different output prefixes differently.
- ffddf34a8: Add new reporting option: `streamLifecycleOutput`. When `true`, the output from child processes is printed immediately and is never collapsed.

### Patch Changes

- Updated dependencies [ffddf34a8]
  - @pnpm/config@9.1.0

## 7.2.5

### Patch Changes

- Updated dependencies [242cf8737]
- Updated dependencies [da091c711]
- Updated dependencies [e11019b89]
- Updated dependencies [802d145fc]
- Updated dependencies [45fdcfde2]
  - @pnpm/config@9.0.0
  - @pnpm/types@6.0.0
  - @pnpm/core-loggers@4.0.2
  - @pnpm/error@1.2.1

## 7.2.5-alpha.2

### Patch Changes

- Updated dependencies [242cf8737]
- Updated dependencies [45fdcfde2]
  - @pnpm/config@9.0.0-alpha.2

## 7.2.5-alpha.1

### Patch Changes

- Updated dependencies [da091c71]
  - @pnpm/types@6.0.0-alpha.0
  - @pnpm/config@8.3.1-alpha.1
  - @pnpm/core-loggers@4.0.2-alpha.0

## 7.2.5-alpha.0

### Patch Changes

- @pnpm/config@8.3.1-alpha.0

## 7.2.4

### Patch Changes

- 907c63a48: Global warnings are reported.
