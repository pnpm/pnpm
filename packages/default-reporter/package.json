{
  "name": "@pnpm/default-reporter",
  "version": "6.1.0",
  "description": "The default reporter of pnpm",
  "main": "lib/index.js",
  "typings": "lib/index.d.ts",
  "files": [
    "lib"
  ],
  "scripts": {
    "lint": "tslint -c ../../tslint.json src/**/*.ts test/**/*.ts",
    "pretty-test": "ts-node test | tap-diff",
    "test": "pnpm run just-test",
    "just-test-preview": "ts-node test --type-check",
    "just-test": "pnpm run tsc && cd ../.. && c8 --reporter lcov --reports-dir packages/default-reporter/coverage ts-node packages/default-reporter/test --type-check",
    "tsc": "rimraf lib && tsc",
    "prepublishOnly": "pnpm run tsc"
  },
  "repository": "https://github.com/pnpm/pnpm/blob/master/packages/default-reporter",
  "keywords": [
    "pnpm-reporter"
  ],
  "author": {
    "name": "Zoltan Kochan",
    "email": "z@kochan.io",
    "url": "https://www.kochan.io/",
    "twitter": "ZoltanKochan"
  },
  "engines": {
    "node": ">=10"
  },
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/pnpm/pnpm/issues"
  },
  "dependencies": {
    "@pnpm/config": "workspace:8.1.1",
    "@pnpm/core-loggers": "workspace:4.0.1",
    "@pnpm/error": "workspace:1.1.0",
    "@pnpm/types": "workspace:5.0.0",
    "@zkochan/zen-push": "1.0.0",
    "ansi-diff": "1.1.1",
    "chalk": "3.0.0",
    "common-tags": "1.8.0",
    "most": "1.8.0",
    "normalize-path": "3.0.0",
    "pretty-bytes": "5.3.0",
    "pretty-time": "1.1.0",
    "ramda": "0.27.0",
    "right-pad": "1.0.1",
    "semver": "7.1.3",
    "stacktracey": "1.2.123",
    "string-length": "4.0.0",
    "strip-ansi": "6.0.0"
  },
  "devDependencies": {
    "@pnpm/default-reporter": "link:",
    "@pnpm/logger": "3.1.0",
    "@types/common-tags": "1.8.0",
    "@types/normalize-path": "3.0.0",
    "@types/pretty-time": "1.1.0",
    "@types/ramda": "^0.26.43",
    "@types/semver": "^7.1.0",
    "delay": "4.3.0",
    "ghooks": "2.0.4",
    "load-json-file": "6.2.0",
    "normalize-newline": "3.0.0"
  }
}
