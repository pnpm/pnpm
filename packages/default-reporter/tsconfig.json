{
  "extends": "@pnpm/tsconfig",
  "compilerOptions": {
    "outDir": "lib",
    "rootDir": "src"
  },
  "include": [
    "src/**/*.ts",
    "../../typings/**/*.d.ts"
  ],
  "references": [
    {
      "path": "../config"
    },
    {
      "path": "../core-loggers"
    },
    {
      "path": "../error"
    },
    {
      "path": "../types"
    }
  ]
}
