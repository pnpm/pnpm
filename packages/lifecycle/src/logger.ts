import baseLogger, {
  LogBase,
  Logger,
} from '@pnpm/logger'

export const lifecycleLogger = baseLogger('lifecycle') as Logger<LifecycleMessage>

// TODO: make depPath optional
export type LifecycleMessage = {
  depPath: string,
  stage: string,
  wd: string,
} & ({
  line: string,
  stdio: 'stdout' | 'stderr',
} | {
  exitCode: number,
} | {
  script: string,
})

export type LifecycleLog = {name: 'pnpm:lifecycle'} & LogBase & LifecycleMessage
