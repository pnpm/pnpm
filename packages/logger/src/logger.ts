import bole from 'bole'

bole.setFastTime()

export const logger = bole('pnpm') as Logger<object>

export interface Logger<T> {
  <Y>(name: string): Logger<Y>
  debug: (log?: T) => void
  info: (log: { message: string, prefix: string }) => void
  warn: (log: { message: string, prefix: string, error?: Error }) => void
  error: (err: Error, log?: string | Error) => void
}

const globalLogger = bole('pnpm:global')

export function globalWarn (message: string): void {
  globalLogger.warn(message)
}

export function globalInfo (message: string): void {
  globalLogger.info(message)
}
