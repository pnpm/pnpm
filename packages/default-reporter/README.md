# @pnpm/default-reporter

> The default reporter of pnpm

## Installation

```
pnpm add @pnpm/default-reporter
```

## Usage

```ts
import { streamParser } from '@pnpm/logger'
import { initDefaultReporter } from '@pnpm/default-reporter'

const stopReporting = initDefaultReporter({
  context: {
    argv: [],
  },
  streamParser,
})

try {
  // calling some pnpm APIs
} finally {
  stopReporting()
}
```

## Style Guide

1. Never use blue or grey as font color as they are hard to read in many consoles.
   1. Use dim instead of grey
   1. Use cyan bright instead of blue
1. Don't hide the CLI cursor. (It is easier to never hide but it is really needed only when scripts are running.)
1. Don't use green and yellow to distinct something.

## License

[MIT](LICENSE)
