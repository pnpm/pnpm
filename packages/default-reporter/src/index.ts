import { Config } from '@pnpm/config'
import * as logs from '@pnpm/core-loggers'
import { LogLevel } from '@pnpm/logger'
import PushStream from '@zkochan/zen-push'
import createDiffer = require('ansi-diff')
import most = require('most')
import { EOL } from './constants'
import mergeOutputs from './mergeOutputs'
import reporterForClient from './reporterForClient'
import reporterForServer from './reporterForServer'

export default function (
  opts: {
    streamParser: object,
    reportingOptions?: {
      appendOnly?: boolean,
      logLevel?: LogLevel,
      throttleProgress?: number,
      outputMaxWidth?: number,
    },
    context: {
      argv: string[],
      config?: Config,
    },
  },
) {
  if (opts.context.argv[0] === 'server') {
    const log$ = most.fromEvent<logs.Log>('data', opts.streamParser)
    reporterForServer(log$)
    return
  }
  const outputMaxWidth = opts.reportingOptions?.outputMaxWidth ?? (process.stdout.columns && process.stdout.columns - 2) ?? 80
  const output$ = toOutput$({ ...opts, reportingOptions: { ...opts.reportingOptions, outputMaxWidth } })
  if (opts.reportingOptions?.appendOnly) {
    output$
      .subscribe({
        complete () {}, // tslint:disable-line:no-empty
        error: (err) => console.error(err.message),
        next: (line) => console.log(line),
      })
    return
  }
  const diff = createDiffer({
    height: process.stdout.rows,
    outputMaxWidth,
  })
  output$
    .subscribe({
      complete () {}, // tslint:disable-line:no-empty
      error: (err) => logUpdate(err.message),
      next: logUpdate,
    })
  function logUpdate (view: string) {
    // A new line should always be appended in case a prompt needs to appear.
    // Without a new line the prompt will be joined with the previous output.
    // An example of such prompt may be seen by running: pnpm update --interactive
    if (!view.endsWith(EOL)) view += EOL
    process.stdout.write(diff.update(view))
  }
}

export function toOutput$ (
  opts: {
    streamParser: object,
    reportingOptions?: {
      appendOnly?: boolean,
      logLevel?: LogLevel,
      outputMaxWidth?: number,
      throttleProgress?: number,
    },
    context: {
      argv: string[],
      config?: Config,
    },
  },
): most.Stream<string> {
  opts = opts || {}
  const fetchingProgressPushStream = new PushStream()
  const progressPushStream = new PushStream()
  const stagePushStream = new PushStream()
  const deprecationPushStream = new PushStream()
  const summaryPushStream = new PushStream()
  const lifecyclePushStream = new PushStream()
  const statsPushStream = new PushStream()
  const installCheckPushStream = new PushStream()
  const registryPushStream = new PushStream()
  const rootPushStream = new PushStream()
  const packageManifestPushStream = new PushStream()
  const linkPushStream = new PushStream()
  const otherPushStream = new PushStream()
  const hookPushStream = new PushStream()
  const skippedOptionalDependencyPushStream = new PushStream()
  const scopePushStream = new PushStream()
  setTimeout(() => { // setTimeout is a workaround for a strange bug in most https://github.com/cujojs/most/issues/491
    opts.streamParser['on']('data', (log: logs.Log) => {
      switch (log.name) {
        case 'pnpm:fetching-progress':
          fetchingProgressPushStream.next(log)
          break
        case 'pnpm:progress':
          progressPushStream.next(log)
          break
        case 'pnpm:stage':
          stagePushStream.next(log)
          break
        case 'pnpm:deprecation':
          deprecationPushStream.next(log)
          break
        case 'pnpm:summary':
          summaryPushStream.next(log)
          break
        case 'pnpm:lifecycle':
          lifecyclePushStream.next(log)
          break
        case 'pnpm:stats':
          statsPushStream.next(log)
          break
        case 'pnpm:install-check':
          installCheckPushStream.next(log)
          break
        case 'pnpm:registry':
          registryPushStream.next(log)
          break
        case 'pnpm:root':
          rootPushStream.next(log)
          break
        case 'pnpm:package-manifest':
          packageManifestPushStream.next(log)
          break
        case 'pnpm:link':
          linkPushStream.next(log)
          break
        case 'pnpm:hook':
          hookPushStream.next(log)
          break
        case 'pnpm:skipped-optional-dependency':
          skippedOptionalDependencyPushStream.next(log)
          break
        case 'pnpm:scope':
          scopePushStream.next(log)
          break
        case 'pnpm' as any: // tslint:disable-line
        case 'pnpm:store' as any: // tslint:disable-line
        case 'pnpm:lockfile' as any: // tslint:disable-line
          otherPushStream.next(log)
          break
      }
    })
  }, 0)
  const log$ = {
    deprecation: most.from<logs.DeprecationLog>(deprecationPushStream.observable),
    fetchingProgress: most.from<logs.FetchingProgressLog>(fetchingProgressPushStream.observable),
    hook: most.from<logs.HookLog>(hookPushStream.observable),
    installCheck: most.from<logs.InstallCheckLog>(installCheckPushStream.observable),
    lifecycle: most.from<logs.LifecycleLog>(lifecyclePushStream.observable),
    link: most.from<logs.LinkLog>(linkPushStream.observable),
    other: most.from<logs.Log>(otherPushStream.observable),
    packageManifest: most.from<logs.PackageManifestLog>(packageManifestPushStream.observable),
    progress: most.from<logs.ProgressLog>(progressPushStream.observable),
    registry: most.from<logs.RegistryLog>(registryPushStream.observable),
    root: most.from<logs.RootLog>(rootPushStream.observable),
    scope: most.from<logs.ScopeLog>(scopePushStream.observable),
    skippedOptionalDependency: most.from<logs.SkippedOptionalDependencyLog>(skippedOptionalDependencyPushStream.observable),
    stage: most.from<logs.StageLog>(stagePushStream.observable),
    stats: most.from<logs.StatsLog>(statsPushStream.observable),
    summary: most.from<logs.SummaryLog>(summaryPushStream.observable),
  }
  const outputs: Array<most.Stream<most.Stream<{msg: string}>>> = reporterForClient(
    log$,
    {
      appendOnly: opts.reportingOptions?.appendOnly,
      cmd: opts.context.argv[0],
      isRecursive: opts.context.config?.['recursive'] === true,
      logLevel: opts.reportingOptions?.logLevel,
      pnpmConfig: opts.context.config,
      throttleProgress: opts.reportingOptions?.throttleProgress,
      width: opts.reportingOptions?.outputMaxWidth,
    },
  )

  if (opts.reportingOptions?.appendOnly) {
    return most.join(
      most.mergeArray(outputs)
      .map((log: most.Stream<{msg: string}>) => log.map((msg) => msg.msg)),
    )
  }
  return mergeOutputs(outputs).multicast()
}
