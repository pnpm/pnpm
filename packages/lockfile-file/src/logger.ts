import { logger } from '@pnpm/logger'

export default logger('lockfile')
