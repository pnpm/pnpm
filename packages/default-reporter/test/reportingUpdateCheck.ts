import { Config } from '@pnpm/config'
import { updateCheckLogger } from '@pnpm/core-loggers'
import { toOutput$ } from '@pnpm/default-reporter'
import { createStreamParser } from '@pnpm/logger'
import { take } from 'rxjs/operators'
import stripAnsi from 'strip-ansi'

test('does not print update if latest is less than current', (done) => {
  const output$ = toOutput$({
    context: {
      argv: ['install'],
    },
    streamParser: createStreamParser(),
  })

  updateCheckLogger.debug({
    currentVersion: '10.0.0',
    latestVersion: '9.0.0',
  })

  const subscription = output$.subscribe({
    complete: () => done(),
    error: done,
    next: () => {
      done('should not log anything')
    },
  })

  setTimeout(() => {
    done()
    subscription.unsubscribe()
  }, 10)
})

test('print update notification if the latest version is greater than the current', (done) => {
  const output$ = toOutput$({
    context: {
      argv: ['install'],
      config: { recursive: true } as Config,
    },
    streamParser: createStreamParser(),
  })

  updateCheckLogger.debug({
    currentVersion: '10.0.0',
    latestVersion: '11.0.0',
  })

  expect.assertions(1)

  output$.pipe(take(1)).subscribe({
    complete: () => done(),
    error: done,
    next: output => {
      expect(stripAnsi(output)).toMatchSnapshot()
    },
  })
})
