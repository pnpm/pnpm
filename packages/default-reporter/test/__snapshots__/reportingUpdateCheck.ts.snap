// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`print update notification if the latest version is greater than the current 1`] = `
"
   ╭──────────────────────────────────────────────────────────────────╮
   │                                                                  │
   │                Update available! 10.0.0 → 11.0.0.                │
   │   Changelog: https://github.com/pnpm/pnpm/releases/tag/v11.0.0   │
   │                 Run pnpm add -g pnpm to update.                  │
   │                                                                  │
   │      Follow @pnpmjs for updates: https://twitter.com/pnpmjs      │
   │                                                                  │
   ╰──────────────────────────────────────────────────────────────────╯
"
`;
